python/pyspark/ml/functions.py

```diff
@@ -106,6 +138,605 @@ def array_to_vector(col: Column) -> Column:
+    # +--------------------+------------------+
+
+    def multi_sum_fn():
+        def predict(x1: np.ndarray, x2: np.ndarray) -> np.ndarray:
+            # x1.shape = [batch_size, 4]
+            # x2.shape = [batch_size, 3]
+            # outputs.shape = [batch_size], result_type = FloatType()
+            return np.sum(x1, axis=1) + np.sum(x2, axis=1)
+
+        return predict
+
+    sum_cols = predict_batch_udf(
+        multi_sum_fn,
+        return_type=FloatType(),
+        batch_size=5,
+        input_tensor_shapes=[[4], [3]],
+    )
+
+    df.withColumn("sum", sum_cols("t1", "t2")).show(5)
+    # +--------------------+------------------+-----+
+    # |                  t1|                t2|  sum|
+    # +--------------------+------------------+-----+
+    # |[0.0, 1.0, 2.0, 3.0]|   [0.0, 1.0, 2.0]|  9.0|
+    # |[4.0, 5.0, 6.0, 7.0]|   [4.0, 5.0, 6.0]| 37.0|
+    # |[8.0, 9.0, 10.0, ...|  [8.0, 9.0, 10.0]| 65.0|
+    # |[12.0, 13.0, 14.0...|[12.0, 13.0, 14.0]| 93.0|
+    # |[16.0, 17.0, 18.0...|[16.0, 17.0, 18.0]|121.0|
+    # +--------------------+------------------+-----+
+
+    # Note that some models can provide multiple outputs.  These can be returned as a dictionary
+    # of named values, which can be represented in columnar (or row-based) formats.
+
+    def multi_sum_fn():
+        def predict_columnar(x1: np.ndarray, x2: np.ndarray) -> Mapping[str, np.ndarray]:
+            # x1.shape = [batch_size, 4]
+            # x2.shape = [batch_size, 3]
+            return {
+                "sum1": np.sum(x1, axis=1),
+                "sum2": np.sum(x2, axis=1)
+            }  # return_type = StructType()
+
+        return predict_columnar
+
+    sum_cols = predict_batch_udf(
+        multi_sum_fn,
+        return_type=StructType([
+            StructField("sum1", FloatType(), True),
+            StructField("sum2", FloatType(), True)
+        ])
+        batch_size=5,
+        input_tensor_shapes=[[4], [3]],
+    )
+
+    df.withColumn("preds", sum_cols("t1", "t2")).select("t1", "t2", "preds.*").show(5)
+    # +--------------------+------------------+----+----+
+    # |                  t1|                t2|sum1|sum2|
+    # +--------------------+------------------+----+----+
+    # |[0.0, 1.0, 2.0, 3.0]|   [0.0, 1.0, 2.0]| 6.0| 3.0|
+    # |[4.0, 5.0, 6.0, 7.0]|   [4.0, 5.0, 6.0]|22.0|15.0|
+    # |[8.0, 9.0, 10.0, ...|  [8.0, 9.0, 10.0]|38.0|27.0|
+    # |[12.0, 13.0, 14.0...|[12.0, 13.0, 14.0]|54.0|39.0|
+    # |[16.0, 17.0, 18.0...|[16.0, 17.0, 18.0]|70.0|51.0|
+    # +--------------------+------------------+----+----+
+    ```
+
+    .. versionadded:: 3.4.0
+
+    Parameters
+    ----------
+    predict_batch_fn : Callable[[],
+        Callable[..., np.ndarray | Mapping[str, np.ndarray] | List[Mapping[str, np.dtype]] ]
+        Function which is responsible for loading a model and returning a `predict` function which
+        takes one or more numpy arrays as input and returns one of the following:
```

Comment:
Took a further look at this and I'm not sure that we should just adopt the Pandas UDF Series/DataFrame return type, since this pushes the conversion code to the end user, e.g. for MNIST, we currently have something like this:
```
def make_predict_fn():
    # note: everything inside this function looks like fairly standard DL code.
    import tensorflow as tf
    model = tf.keras.models.load_model('/path/to/mnist_model')
    def predict(inputs: np.ndarray) -> np.ndarray:
        return model.predict(inputs)
        
    return predict

mnist = predict_batch_udf(make_predict_fn,
                          return_type=ArrayType(FloatType()),
                          batch_size=1024,
                          input_tensor_shapes=[[784]])
```

By requiring the user to return a Pandas UDF Series/DataFrame output directly, we'd end up with:
```
def make_predict_fn():
    import pandas as pd       # Now have to introduce pandas, where most DL code wouldn't use it
    import tensorflow as tf
    model = tf.keras.models.load_model('/path/to/mnist_model')
    def predict(inputs: np.ndarray) -> pd.Series | pd.DataFrame:
        # Now user has to correctly pick one of these options (depending on the model output and UDF return_type)
        # While this could be learned/documented, it seems much more difficult than above
        return pd.Series(list(model.predict(inputs))
        return pd.Series(np.squeeze(model.predict(inputs))
        return pd.DataFrame(model.predict(inputs))
        
    return predict
```

Reply:
I rethink on this issue,
and I still prefers that `predict_batch_fn` should return a function with Numpy array / dict type. 
because
* in most ML model cases, the model.predict returns a numpy array or a dict of numpy array, so the conversion to pandas dataframe should be spark side code work. We should make `batch_infer_udf` to be easier to user.
* The input arguments of the function returned by predict_batch_fn are also numpy arrays. So for input / output consistency, we should make output type be numpy array

Reply:
Thanks for the discussion! Let's keep the current return type.